azure/scope/cluster.go

```diff
@@ -457,6 +458,15 @@ func (s *ClusterScope) AdditionalTags() infrav1.Tags {
 	return tags
 }
 
+// APIServerLBFrontendPort returns the frontend listen port to use when creating the load balancer.
+func (s *ClusterScope) APIServerLBFrontendPort() int32 {
+	if s.Cluster.Spec.ControlPlaneEndpoint.Port > 0 {
```

Comment:
What is still not 100% clear to me is the relationship between Cluster.Spec.ControlPlaneEndpoint and AzureCluster.Spec.ControlPlaneEndpoint

We set AzureCluster.Spec.ControlPlaneEndpoint in the AzureCluster controller here (to follow the CAPI provider contract): https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/controllers/azurecluster_controller.go#L256-L260

I had opened a discussion in CAPI a while back about moving it to Status since it's basically being used as status right now (and any user value will be overridden) https://github.com/kubernetes-sigs/cluster-api/issues/3715

Reply:
in the CAPI cluster_controller, `Cluster.Spec.ControlPlaneEndpoint` is set to the value of `AzureCluster.Spec.ControlPlaneEndpoint` https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/cluster_controller_phases.go#L184

Reply:
Hmm... I didn't realize to think from this perspective.

My original use case for this change was to provide migration compatibility when converting existing Kubernetes cluster (Giant Swarm flavor in this case) into one managed by CAPZ. The cluster is already reconciled by v1alpha3 CAPI + CAPZ CRs and there's DNS zone managed for the workload cluster.

I was also viewing the `Cluster` CR as main spec for provider independent workload cluster configuration since it also has several important network configuration fields (e.g. Service and POD CIDRs as well as API server listen port).

Now I see that my understanding / interpretation on that is slightly off from the present implementation and interoperability between different controllers.

Do you think this change would still make sense in CAPZ?

As far as I was able to reason & test, this change should be fully compatible with present state, but I do understand that there's the corner case that one might not be able to configure _just_ port because the host must be set as well and something needs to manage that. I'm not sure if that can be fixed in any way right now.

Reply:
> I was also viewing the Cluster CR as main spec for provider independent workload cluster configuration since it also has several important network configuration fields (e.g. Service and POD CIDRs as well as API server listen port).

Right, that's why it's confusing. I think it's a bug that providers don't honor that field, that's why I opened https://github.com/kubernetes-sigs/cluster-api/issues/3715. 

I think as long as it has no effect on current behaviors for those who don't set the port & for existing clusters that have the default port, it should be ok to make this change to unblock your use case. However in that case we should remove the user facing documentation since it's not really a configuration option, but more of a workaround. Also let's make sure there are solid comments in the code so we know exactly why this was done and what the expected bahavior is when we look at this code again 6 months from now.

Reply:
Sorry for long delays, we have been recently working on different areas so this was not high priority. Because of this controversial nature of this feature and the fact that we don't urgently need this right now, I'll take more time to consider this. I'll get back to this later. Thanks for your patience :slightly_smiling_face: